feat: Implement custom function module in milvus expr

jiangyinzuo opened the pull request:
OSPP 2024 project: https://summer-ospp.ac.cn/org/prodetail/247410235?list=org&navpage=org

Solutions:

- parser (planparserv2)
    - add CallExpr in planparserv2/Plan.g4
    - update parser_visitor and show_visitor
- grpc protobuf
    - add CallExpr in plan.proto
- execution (`core/src/exec`)
    - add `CallExpr` `ValueExpr` and `ColumnExpr` (both logical and physical) for function call and function parameters
- function factory (`core/src/exec/expression/function`)
    - create a global hashmap when starting milvus (see server.go)
    - the global hashmap stores function signatures and their function pointers, the CallExpr in execution engine can get the function pointer by function signature.
- custom functions
    - empty(string)
    - starts_with(string, string)
- add cpp/go unittests and E2E tests

closes: #36559

jiangyinzuo commented on internal/core/src/exec/expression/CallExpr.cpp:
```diff
@@ -0,0 +1,52 @@
+        throw std::runtime_error("FilterScalarFunction " + func_sig.toString() +
+                                 " not found. " + this->expr_->ToString());
+    }
+    // eval inputs in function
```
Evaluate the input arguments here, not in function

jiangyinzuo commented on internal/core/src/exec/expression/CallExpr.cpp:
```diff
@@ -0,0 +1,52 @@
+    expression::FilterFunctionRegisterKey func_sig{
+        this->expr_->fun_name(), std::move(func_param_type_list)};
+    auto& factory = expression::FunctionFactory::Instance();
+    auto function = factory.GetFilterFunction(func_sig);
```
Call `factory.GetFilterFunction` at logical expr or parser

jiangyinzuo reviewed (commented):
Some todos

jiangyinzuo commented:
@zhagnlu

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo commented:
/run-cpu-e2e

zhagnlu commented on internal/querynodev2/server.go:
```diff
@@ -356,6 +357,8 @@ func (node *QueryNode) Init() error {
 			return
 		}
 
+		C.InitExecExpressionFunctionFactory()
```
move this interface to InitSegcore

zhagnlu commented on internal/core/src/common/Vector.h:
```diff
@@ -100,28 +112,81 @@ class ColumnVector final : public BaseVector {
     }
 
  private:
+    bool is_bitmap_;
```
this is not a good implment, can just using BitmapVector as a sub class of baseVector

jiangyinzuo replied:
Yes, but I think this PR is large enough. Replace `ColumnVector` with `BitmapVector`  in other Exprs would introduce a large number of code modifications. It's better to implement `BitmapVector` in a new PR.

jiangyinzuo replied:
See #37075

zhagnlu commented on internal/core/src/common/Vector.h:
```diff
@@ -58,18 +59,35 @@ class BaseVector {
 
 using VectorPtr = std::shared_ptr<BaseVector>;
 
+class SimpleVector : public BaseVector {
```
why add this class?

jiangyinzuo replied:
In custom functions like `StartsWithVarchar`, we need a base class for `ColumnVector`, `ConstantVector` (and `BitmapVector` in the future). So I add `SimpleVector` as their base class. `BaseVector` only has `size()` and `type` methods, which is unable to meet the requirements.

```cpp
void
StartsWithVarchar(const RowVector& args, FilterFunctionReturn& result) {
    Assert(args.childrens().size() == 2);
    auto strs = std::dynamic_pointer_cast<SimpleVector>(args.child(0));
    if (strs == nullptr) {
        PanicInfo(ExprInvalid, "invalid vector type");
    }
    Assert(strs->type() == DataType::VARCHAR ||
           strs->type() == DataType::STRING);
    auto prefixes = std::dynamic_pointer_cast<SimpleVector>(args.child(1));
    if (prefixes == nullptr) {
        PanicInfo(ExprInvalid, "invalid vector type");
    }
```

zhagnlu commented on internal/core/src/exec/expression/function/FunctionFactory.cpp:
```diff
@@ -0,0 +1,83 @@
+namespace expression {
+
+std::string
+FilterFunctionRegisterKey::toString() const {
```
ToString()

jiangyinzuo commented:
rerun go-sdk

jiangyinzuo commented:
/run-cpu-e2e

zhagnlu commented on tests/python_client/testcases/test_query.py:
```diff
@@ -5598,3 +5598,22 @@ def test_query_text_match_with_unsupported_tokenizer(self):
+        for mixed_call_expr in ["not empty(varchar) && int64 >= 0", "starts_with(varchar, varchar) && int64 >= 0"]:
+            collection_w.query(mixed_call_expr, output_fields=output_fields,
+                               check_task=CheckTasks.check_query_results, check_items={exp_res: res})
+
```
add a negative case for not supported function

jiangyinzuo replied:
Done. I also added `EmPty` and `EMPTY` for checking case-insensitive function calls

jiangyinzuo commented on internal/core/src/exec/expression/function/impl/StartsWith.cpp:
```diff
@@ -0,0 +1,66 @@
+            auto* str_ptr = reinterpret_cast<std::string*>(
+                strs->RawValueAt(i, sizeof(std::string)));
+            auto* prefix_ptr = reinterpret_cast<std::string*>(
+                prefixes->RawValueAt(i, sizeof(std::string)));
```
`StartsWithVarchar` supports both `starts_with(col1, col2)` and `starts_with(col1, "string literal")`

czs007 commented on internal/core/src/segcore/SegmentChunkReader.h:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    MultipleChunkDataAccessor
+    GetChunkData(DataType data_type,
```
Better to rename as GetChunkDataAccessor?

czs007 commented on internal/core/src/common/Vector.h:
```diff
@@ -100,28 +116,81 @@ class ColumnVector final : public BaseVector {
+    bool is_bitmap_;  // TODO: remove the field after implementing BitmapVector
     FieldDataPtr values_;
-    FieldDataPtr valid_values_;
+    TargetBitmap valid_values_;  // false means the value is null
```
@smellthemoon  please help check this change.

smellthemoon replied:
it's ok to do that, but why not also change the type of values_ to TargetBitmap?

jiangyinzuo replied:
Currently `values_` in `ColumnVector` may be scalar type vectors or bitmap. We can change the type of values_ to TargetBitmap in `BitmapVector` in the future.

jiangyinzuo commented on internal/core/src/exec/expression/CompareExpr.h:
```diff
@@ -113,31 +100,39 @@ class PhyCompareFilterExpr : public Expr {
-        size_per_chunk_ = segment_->size_per_chunk();
-        if (segment_->is_chunked()) {
+        is_left_indexed_ =
+            segment_chunk_reader_.segment_->HasIndex(left_field_);
```
Yes, I have extracted some `segment_`-related codes that are commonly used by `CompareExpr` and `ColumnExpr` into the `class SegmentChunkReader`, but there are still some codes in CompareExpr and ColumnExpr that directly use `segment_`.

zhagnlu commented on internal/parser/planparserv2/parser_visitor.go:
```diff
@@ -594,6 +594,28 @@ func (v *ParserVisitor) getChildColumnInfo(identifier, child antlr.TerminalNode)
+		expr: &planpb.Expr{
+			Expr: &planpb.Expr_CallExpr{
+				CallExpr: &planpb.CallExpr{
+					FunctionName:       functionName,
```
> It is auto-generated by protobuf, `UnaryExpr` also uses CamelCase (`Op` and `Child`)
> 
> ```go
> 	if ctx.GetOp() != nil {
> 		expr = &planpb.Expr{
> 			Expr: &planpb.Expr_UnaryExpr{
> 				UnaryExpr: &planpb.UnaryExpr{
> 					Op:    planpb.UnaryExpr_Not,
> 					Child: expr,
> 				},
> 			},
> 		}
> 	}
> ```

I means the function name that input from user

jiangyinzuo replied:
OK. Now `functionName` is case-insensitive

jiangyinzuo commented:
rerun cpp-unit-test

jiangyinzuo commented:
rerun cpp-unit-test

jiangyinzuo reviewed (commented):
I have extracted another 3 methods from `CompareExpr` and `ColumnExpr`.

jiangyinzuo commented on internal/core/src/segcore/SegmentChunkReader.h:
```diff
@@ -0,0 +1,157 @@
+                         int data_barrier) const;
+
+    void
+    MoveCursorForMultipleChunk(int64_t& current_chunk_id,
```
I have extracted another 3 methods from `ColumnExpr` and `CompareExpr`

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo commented:
rerun cpp-unit-test

jiangyinzuo commented:
/run-cpu-e2e

jiangyinzuo committed:
feat: Implement custom function module in milvus expr

OSPP 2024 project: https://summer-ospp.ac.cn/org/prodetail/247410235?list=org&navpage=org

Solutions:

- parser (planparserv2)
    - add CallExpr in planparserv2/Plan.g4
    - update parser_visitor and show_visitor
- grpc protobuf
    - add CallExpr in plan.proto
- execution (`core/src/exec`)
    - add `CallExpr`, `ValueExpr` and `ColumnExpr` (both logical and
      physical) for function call and function parameters
- function factory (`core/src/exec/expression/function`)
    - create a global hashmap when starting milvus (see server.go)
    - the global hashmap stores function signatures and their function
      pointers, the CallExpr in execution engine can get the function pointer by
      function signature.
- `common/Vector.h`
    - add ConstantVector, update ColumnVector
- `segcore/SegmentChunkReader.h`
    - extracted some methods from `CompareExpr.h`
- custom functions
    - empty(string)
    - starts_with(string, string)
- add cpp/go unittests and E2E tests

closes: #36559

Signed-off-by: Yinzuo Jiang <jiangyinzuo@foxmail.com>

jiangyinzuo commented:
rerun cpp-unit-test

jiangyinzuo commented:
rerun cpp-unit-test

jiangyinzuo commented:
rerun cpp-unit-test

czs007 commented:
/approve
